src/event/event_windows.c

```diff
@@ -48,34 +60,193 @@ _dispatch_unote_unregister(dispatch_unote_t du DISPATCH_UNUSED,
-		dispatch_timer_delay_s range DISPATCH_UNUSED,
+_dispatch_event_loop_timer_arm(uint32_t tidx, dispatch_timer_delay_s range,
 		dispatch_clock_now_cache_t nows DISPATCH_UNUSED)
 {
-	WIN_PORT_ERROR();
+	dispatch_windows_timeout_t timer;
+	FILETIME ftDueTime;
+	LARGE_INTEGER liTime;
+
+	switch (DISPATCH_TIMER_CLOCK(tidx)) {
+	case DISPATCH_CLOCK_WALL:
+		timer = &_dispatch_windows_timeout[DISPATCH_CLOCK_WALL];
+
+		WIN_PORT_ERROR();
+		__assume(0);
+
+//	case DISPATCH_CLOCK_UPTIME:
+//		timer = &_dispatch_windows_timeout[DISPATCH_CLOCK_UPTIME];
+//		liTime.QuadPart = -((range.delay + 99) / 100);
+//		break;
+
+	case DISPATCH_CLOCK_MONOTONIC:
+		timer = &_dispatch_windows_timeout[DISPATCH_CLOCK_MONOTONIC];
+
+		liTime.QuadPart = -((range.delay + 99) / 100);
+		break;
+	}
+
+	if (timer->pTimer == NULL) {
+		timer->pTimer = CreateThreadpoolTimer(_dispatch_timer_callback,
+			(LPVOID)timer->ullIdent, NULL);
+		if (timer->pTimer == NULL) {
+			DISPATCH_INTERNAL_CRASH(GetLastError(),
+				"CreateThreadpoolTimer");
+		}
+	}
+
+	ftDueTime.dwHighDateTime = liTime.HighPart;
+	ftDueTime.dwLowDateTime = liTime.LowPart;
+
+	SetThreadpoolTimer(timer->pTimer, &ftDueTime, /*msPeriod=*/0,
+		/*msWindowLength=*/0);
+	timer->bArmed = TRUE;
+
+	_dispatch_timers_heap[tidx].dth_flags |= DTH_ARMED;
 }
 
 void
 _dispatch_event_loop_timer_delete(uint32_t tidx DISPATCH_UNUSED)
 {
-	WIN_PORT_ERROR();
+	dispatch_windows_timeout_t timer;
+
+	switch (DISPATCH_TIMER_CLOCK(tidx)) {
+	case DISPATCH_CLOCK_WALL:
+		timer = &_dispatch_windows_timeout[DISPATCH_CLOCK_WALL];
+		break;
+
+//	case DISPATCH_CLOCK_UPTIME:
+//		timer = &_dispatch_windows_timeout[DISPATCH_CLOCK_UPTIME];
+//		break;
+
+	case DISPATCH_CLOCK_MONOTONIC:
+		timer = &_dispatch_windows_timeout[DISPATCH_CLOCK_MONOTONIC];
+		break;
+	}
+
+	SetThreadpoolTimer(timer->pTimer, NULL, /*msPeriod=*/0,
+		/*msWindowLength=*/0);
+	timer->bArmed = FALSE;
+
+	_dispatch_timers_heap[tidx].dth_flags &= ~DTH_ARMED;
 }
 
 #pragma mark dispatch_loop
 
+static void
+_dispatch_windows_port_init(void *context DISPATCH_UNUSED)
+{
+	hPort = CreateIoCompletionPort(INVALID_HANDLE_VALUE, NULL, 0, 1);
+	if (hPort == NULL) {
+		DISPATCH_INTERNAL_CRASH(GetLastError(), "CreateIoCompletionPort");
+	}
+
+#if DISPATCH_USE_MGR_THREAD
+//	_dispatch_trace_item_push(_dispatch_mgr_q.do_targetq, &_dispatch_mgr_q);
+	dx_push(_dispatch_mgr_q.do_targetq, &_dispatch_mgr_q, 0);
+#endif
+}
+
 void
 _dispatch_event_loop_poke(dispatch_wlh_t wlh DISPATCH_UNUSED,
 		uint64_t dq_state DISPATCH_UNUSED, uint32_t flags DISPATCH_UNUSED)
 {
-	WIN_PORT_ERROR();
+	static dispatch_once_t _dispatch_windows_port_init_pred;
+	BOOL bSuccess;
+
+	dispatch_once_f(&_dispatch_windows_port_init_pred, NULL,
+		_dispatch_windows_port_init);
+	bSuccess = PostQueuedCompletionStatus(hPort, 0, DISPATCH_PORT_POKE,
+		NULL);
+	(void)dispatch_assume(bSuccess);
 }
 
 DISPATCH_NOINLINE
 void
 _dispatch_event_loop_drain(uint32_t flags DISPATCH_UNUSED)
```

Comment:
I think we should remove the `DISPATCH_UNUSED` here because we check the `KEVENT_FLAG_IMMEDIATE` flag.

Reply:
Well, that would just be asking that I learn to write code ... *sigh*.  Thanks